File: backend/internal/availability/availability.go
package availability

import (
	"errors"
	"strconv"
	"strings"

	"courtflow/internal/models"

	"gorm.io/gorm"
)

// toMinutes parses "HH:mm" -> minutes since midnight. Returns -1 on bad input.
func toMinutes(hhmm string) int {
	parts := strings.SplitN(hhmm, ":", 2)
	if len(parts) != 2 {
		return -1
	}
	h, err1 := strconv.Atoi(parts[0])
	m, err2 := strconv.Atoi(parts[1])
	if err1 != nil || err2 != nil {
		return -1
	}
	return h*60 + m
}

// Overlap implements PRD §13: existing.start < new.end AND existing.end > new.start.
func Overlap(aStart, aEnd, bStart, bEnd string) bool {
	as, ae := toMinutes(aStart), toMinutes(aEnd)
	bs, be := toMinutes(bStart), toMinutes(bEnd)
	return as < be && ae > bs
}

// DurationHours returns the booking length in hours (may be fractional).
func DurationHours(start, end string) float64 {
	return float64(toMinutes(end)-toMinutes(start)) / 60.0
}

// Valid checks start<end and both parse.
func ValidRange(start, end string) bool {
	s, e := toMinutes(start), toMinutes(end)
	return s >= 0 && e >= 0 && e > s
}

type Slot struct {
	StartTime string `json:"startTime"`
	EndTime   string `json:"endTime"`
	Available bool   `json:"available"`
	Reason    string `json:"reason,omitempty"` // booked | blocked | outside_hours
}

// weekday: 0=Sun..6=Sat. date is YYYY-MM-DD.
// Generates hourly slots within operating hours, marking each unavailable
// if outside hours, blocked, or overlapped by an active booking (AC-4.6..4.10).
func Compute(db *gorm.DB, courtID uint, date string, weekday int) ([]Slot, error) {
	var oh models.OperatingHour
	err := db.Where("court_id = ? AND day_of_week = ?", courtID, weekday).First(&oh).Error
	if errors.Is(err, gorm.ErrRecordNotFound) || (err == nil && oh.Closed) {
		return []Slot{}, nil // closed/unscheduled day -> no slots (AC-4.3)
	}
	if err != nil {
		return nil, err // real DB error: do not pretend slots are available
	}

	var blocked []models.BlockedTime
	if err := db.Where("court_id = ? AND date = ?", courtID, date).Find(&blocked).Error; err != nil {
		return nil, err
	}

	var active []models.Booking
	if err := db.Where("court_id = ? AND date = ? AND status IN ?", courtID, date,
		[]string{models.BookingPendingPayment, models.BookingConfirmed}).Find(&active).Error; err != nil {
		return nil, err
	}

	open, close := toMinutes(oh.OpenTime), toMinutes(oh.CloseTime)
	slots := []Slot{}
	for t := open; t+60 <= close; t += 60 {
		start := fmtMin(t)
		end := fmtMin(t + 60)
		slot := Slot{StartTime: start, EndTime: end, Available: true}

		for _, b := range blocked {
			if Overlap(b.StartTime, b.EndTime, start, end) {
				slot.Available, slot.Reason = false, "blocked"
				break
			}
		}
		if slot.Available {
			for _, b := range active {
				if Overlap(b.StartTime, b.EndTime, start, end) {
					slot.Available, slot.Reason = false, "booked"
					break
				}
			}
		}
		slots = append(slots, slot)
	}
	return slots, nil
}

func fmtMin(m int) string {
	h := m / 60
	mm := m % 60
	return pad(h) + ":" + pad(mm)
}

func pad(n int) string {
	if n < 10 {
		return "0" + strconv.Itoa(n)
	}
	return strconv.Itoa(n)
}

File: frontend/src/lib/availability.ts
// Slot/availability calculation (PRD §4-5). Pure helpers for UI-side checks.
import type { Booking, BlockedTime, OperatingHour, Slot } from "./types";
import { overlaps } from "./format";

const ACTIVE: Booking["status"][] = ["pending_payment", "confirmed"];

// Build hourly slots for a court+date from operating hours, blocked time, active bookings.
export function buildSlots(
  courtId: number,
  date: string,
  hours: OperatingHour[],
  blocked: BlockedTime[],
  bookings: Booking[],
): Slot[] {
  const dow = new Date(`${date}T00:00:00+07:00`).getDay();
  const oh = hours.find((h) => h.courtId === courtId && h.dayOfWeek === dow);
  if (!oh || oh.closed) return [];

  const open = parseInt(oh.openTime.slice(0, 2), 10);
  const close = parseInt(oh.closeTime.slice(0, 2), 10);

  const dayBlocked = blocked.filter((b) => b.courtId === courtId && b.date === date);
  const dayBookings = bookings.filter(
    (b) => b.courtId === courtId && b.date === date && ACTIVE.includes(b.status),
  );

  const slots: Slot[] = [];
  for (let h = open; h < close; h++) {
    const start = `${String(h).padStart(2, "0")}:00`;
    const end = `${String(h + 1).padStart(2, "0")}:00`;

    const isBooked = dayBookings.some((b) => overlaps(b.startTime, b.endTime, start, end));
    const isBlocked = dayBlocked.some((b) => overlaps(b.startTime, b.endTime, start, end));

    slots.push({
      startTime: start,
      endTime: end,
      available: !isBooked && !isBlocked,
      reason: isBooked ? "booked" : isBlocked ? "blocked" : undefined,
    });
  }
  return slots;
}

// Backend-style final check (PRD AC-6.6): does a proposed range collide?
export function hasConflict(
  courtId: number,
  date: string,
  start: string,
  end: string,
  blocked: BlockedTime[],
  bookings: Booking[],
): boolean {
  const b = blocked.some(
    (x) => x.courtId === courtId && x.date === date && overlaps(x.startTime, x.endTime, start, end),
  );
  const k = bookings.some(
    (x) =>
      x.courtId === courtId &&
      x.date === date &&
      ACTIVE.includes(x.status) &&
      overlaps(x.startTime, x.endTime, start, end),
  );
  return b || k;
}
